### PR TITLE
Implement Send Invoice flow from Global Create

### PR DESCRIPTION
**Pull request by VickyStash**
### Details

Implementation of invoice creation flow from Global Create

### Fixed Issues

$ https://github.com/Expensify/App/issues/40012
PROPOSAL: N/A

### Tests

- [x] Verify that no errors appear in the JS console

**Note (!):** Use staging server by turning it on in `Settings` -> `Troubleshoot` -> `Use Staging Server`
**Note (!):** the backend has a know [issue](https://github.com/Expensify/App/issues/41068) related to the invoice currency. If you want your test results to not be affected by that issue, set the sender workspace and the invoice currencies to be USD.

**Flow 1**

1. Sign in to the app with a workspace admin account
2.  Open Global Create and verify `Send Invoice` appears in Global Create
3.  Click Send invoice in Global Create:
3.1 Verify “Distance” and “Scan" do not appear in the big number pad (BNP) screen.
3.2 Verify the Currency can be modified
4. Add an amount  -> tap Next
5. Select the participant, and make sure the participant's step works as expected. Press Next.
6. Verify all of the information on the confirmation screen looks as expected.
6.1 The sender Workspace is displayed as expected.
6.2 The receiver is displayed as expected.
7. Tap on the `Send from` workspace, you should be navigated to the Send from the screen.
8. Select another workspace, and make sure it works as expected.
9.  tap Send $XX.XX invoice, verify all of the necessary data is created in Onyx.

**Flow 2**
1. Repeat Flow 1, but select the same sender and receiver.
2. Make sure you are navigated to the same room after invoice confirmation.

**Flow 3**
1. Sign in to the app with an account without workspaces
3. Confirm you have no `Send Invoice` option in Global Create

### Offline tests

Same as in the **Tests** section

### QA Steps

- [x] Verify that no errors appear in the JS console

Same as in the **Tests** section

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/23176449/6b690c1a-9bdf-4f0b-8ef8-746b09e688c3

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/23176449/7e2e98ba-64b4-47a9-a6cf-008097071c2e

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/23176449/9aed0d1c-527b-4d61-a6b7-c9506621f751

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/23176449/9f0479cb-cc97-450c-9752-55e755d730fb

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/23176449/dabd2cec-add8-478b-816a-11079abd3e62

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/23176449/adb5e566-a501-4531-9ae8-fd5dc17b0140

</details>

### CODE REVIEW

File: src/languages/es.ts
Comment (VickyStash) [inline src/languages/es.ts:1]: Need the confirmation regarding this translations

File: src/components/MoneyTemporaryForRefactorRequestConfirmationList.tsx
Comment (rezkiy37) [inline src/components/MoneyTemporaryForRefactorRequestConfirmationList.tsx:256]:
```suggestion
        () => PolicyUtils.canSendInvoice(allPolicies) > 0 && !!transaction?.isFromGlobalCreate,
```

File: src/libs/actions/IOU.ts
Comment (rezkiy37) [inline src/libs/actions/IOU.ts:826]: Can we use an object instead of these arguments?

File: src/libs/actions/IOU.ts
Comment (rezkiy37) [inline src/libs/actions/IOU.ts:846]: If something we can add without any conditions, let's set it by default to avoid an extra `push`.

File: src/libs/actions/IOU.ts
Comment (rezkiy37) [inline src/libs/actions/IOU.ts:948]: If something we can add without any conditions, let's set it by default to avoid an extra `push`.

File: src/libs/actions/IOU.ts
Comment (rezkiy37) [inline src/libs/actions/IOU.ts:1416]: An object instead of many arguments?

File: src/libs/actions/IOU.ts
Comment (rezkiy37) [inline src/libs/actions/IOU.ts:2793]: An object instead of many arguments?

**Review by rezkiy37 — CHANGES_REQUESTED** [review]
Left comments.

File: src/libs/actions/IOU.ts
Comment (VickyStash) [reply src/libs/actions/IOU.ts]: It's the same way in other places in this file, so I decided to keep it consistent. Let's see how further review will go, I'll do quick update if it's necessary

File: src/libs/actions/IOU.ts
Comment (VickyStash) [reply src/libs/actions/IOU.ts:3328]: It's the same way in other places in this file, so I decided to keep it consistent. Let's see how further review will go, I'll do quick update if it's necessary

**Comment by VickyStash** [comment]
**Updates:**
- Updated param invoiceRoomID -> invoiceRoomReportID
- Updates to follow main branch changes (MoneyTemporaryForRefactorRequestConfirmationList.tsx got removed)
- Tested API updates

**TODOs:**
- Waiting for API fixes to be able to progress PR

**Comment by VickyStash** [comment]
This PR is still waiting for some API fixes, but due to the short timeline, we need to start the review process the sooner the better!
cc @cristipaval

**Comment by shawnborton** [comment]
For the big number pad screen, we decided that there should be no tab control at the top:
CleanShot 2024-04-24 at 11 01 22@2x

**Comment by shawnborton** [comment]
Also just wanted to double check with @JmillsExpensify @dannymcclain @davidcardoza - are we going to add the Next steps right into the preview card? Looks like this PR has them:
CleanShot 2024-04-24 at 11 03 22@2x

I thought we might have decided to punt that?

**Comment by VickyStash** [comment]
@shawnborton Unfortunately, there are some API known issues that block me from doing the correct  screenshots/videos without the offline mode. But it should be resolved tomorrow morning, and I'll update the recordings as soon as it will be done

**Comment by shawnborton** [comment]
Sounds good!

**Comment by dannymcclain** [comment]
>Also just wanted to double check with @JmillsExpensify @dannymcclain @davidcardoza - are we going to add the Next steps right into the preview card?
>I thought we might have decided to punt that?

Yes I believe we did decide to not tackle that right now.

**Comment by VickyStash** [comment]
@shawnborton @dannymcclain 

> Yes I believe we did decide to not tackle that right now.

Just to confirm, we shouldn't show `Awaiting payment by xxx` for now, right?

^^ @rezkiy37

**Comment by shawnborton** [comment]
Let's get confirmation from @JmillsExpensify on that first. If we implement it for invoices, we should be implementing it for all preview cards as well.

**Comment by davidcardoza** [comment]
Brought the topic up in the VIP room - https://expensify.slack.com/archives/CSL3XBCCR/p1713994340869369

File: assets/images/invoice-generic.svg
Comment (shawnborton) [reply assets/images/invoice-generic.svg:4]: We usually supply them at pure black (`#000000`) - maybe we just need to edit the fill value here as a quick fix? Otherwise I am happy to provide a new asset.

File: src/pages/iou/request/step/IOURequestStepSendFrom.tsx
Comment (shawnborton) [reply src/pages/iou/request/step/IOURequestStepSendFrom.tsx:93]: Can you share a screenshot of what you are referring to?

I think I am following you though - I suppose we could make the selection look more like this screen?

CleanShot 2024-04-25 at 21 29 05@2x

Comment (shawnborton) [reply src/pages/iou/request/step/IOURequestStepSendFrom.tsx:93]: Got it I don't feel strongly, I would be totally down to use the other list style I referenced above.

File: assets/images/invoice-generic.svg
Comment (shawnborton) [reply assets/images/invoice-generic.svg:4]: Here is an update, exported the same exact way I've exported many other icons in the repo. 
[invoice-generic.svg.zip](https://github.com/Expensify/App/files/15124398/invoice-generic.svg.zip)

Comment (VickyStash) [reply assets/images/invoice-generic.svg]: Updated to use provided icon - https://github.com/Expensify/App/pull/40015/commits/e22c3eb565a51ad21ddc94eaaba237cce3358a2d

File: src/pages/iou/request/step/IOURequestStepSendFrom.tsx
Comment (VickyStash) [reply src/pages/iou/request/step/IOURequestStepSendFrom.tsx]: Updated to use default SectionList checkmark - https://github.com/Expensify/App/pull/40015/commits/f3d02339a9c32f2d893660effa4753ae3dd4d861
image

**Comment by VickyStash** [comment]
@shubham1206agra I've applied your feedback, please take another look!
I've also updated Notes in the PR **Tests** section, so please take a look at them before testing.
Updated recordings in online mode have also been attached.

File: src/pages/iou/request/step/IOURequestStepSendFrom.tsx
Comment (dannymcclain) [reply src/pages/iou/request/step/IOURequestStepSendFrom.tsx]: Good shout. I think this should just use our regular select list styles—not multi-select. Examples below.

CleanShot 2024-04-26 at 08 33 36@2x

CleanShot 2024-04-26 at 08 35 07@2x

CleanShot 2024-04-26 at 08 35 57@2x

Comment (VickyStash) [reply src/pages/iou/request/step/IOURequestStepSendFrom.tsx]: @dannymcclain Updated a little bit more to follow your screenshots 🙂 
image

**Comment by cristipaval** [comment]
@shawnborton @dannymcclain is this PR approved from the design standpoint?

**Comment by cristipaval** [comment]
> > the display of Invoice room and related report actions is implemented in this PR #40303. So until the PR is not merged, you need to pull it manually to this for testing.
> 
> @VickyStash Do I need to do anything extra or changes are already present here?

@shubham1206agra I held this one on https://github.com/Expensify/App/pull/40303

**Comment by VickyStash** [comment]
> @VickyStash Do I need to do anything extra or changes are already present here?

@shubham1206agra You don't need to make any additional changes, maybe just minor conflict resolution can be required

**Comment by cristipaval** [comment]
Wooot! 🎉  [The other PR](https://github.com/Expensify/App/pull/40303) is merged. Let's pull `main` into this one and test the ent-to-end Send Invoice flow

**Comment by VickyStash** [comment]
@shubham1206agra I've updated the PR, please take a look 👀

**Comment by dubielzyk-expensify** [comment]
Looks good at my first glance, but I know @shawnborton is closer to this, so I'll let him chime in

**Comment by VickyStash** [comment]
@shubham1206agra 

> And why did we allowed addition of receipts in this?

Could you please clarify where were you able to ass a receipt?

**Comment by VickyStash** [comment]
> The subtitle seems to be broken.

@shubham1206agra Could you please let me know what is the expected subtitle?

**Comment by VickyStash** [comment]
> See 2 `Shubham:` written

Sorry, I don't get it, could you please clarify

**Comment by cristipaval** [comment]
> @VickyStash Erroneous total and GBR showing while editing amount in invoice room. cc @cristipaval
@shubham1206agra the wrong amount after transaction edit is due to a backend issue that we fixed but it is not deployed yet. We are aware of it so please do not block this PR on this issue 🙏

**Comment by VickyStash** [comment]
> No GBR present here.

@shubham1206agra GBR and Pay button are out of scope, I've hidden Pay button as well in this commit https://github.com/Expensify/App/pull/40015/commits/c36bd3bec9a56c42b628ed7d188cbfd85061c4f6

**Comment by VickyStash** [comment]
> Screenshot 2024-04-29 at 11 59 00 AM Screenshot 2024-04-29 at 11 59 23 AM
> After doing `clear cache and restart`, the App crashes on opening invoices. (I have added a receipt in one invoice).

@shubham1206agra I've made some investigations and it looks like it can be related to another issue. There is an open [issue](https://github.com/Expensify/App/issues/41130) with the app crashing with another flow, but the same error. There is a [proposal](https://github.com/Expensify/App/issues/41130#issuecomment-2080881543) that describes the reason of the problem and the solution, so I'm not sure if I should resolve it in the current PR

**Comment by VickyStash** [comment]
> @VickyStash One minor thing I noticed is that QAB gets updated on Send Invoice. This shouldn't happen.

@shubham1206agra Thank you, fixed - https://github.com/Expensify/App/pull/40015/commits/43710829df28df950ce6b2364ccfade09f4831b3

**Review by shubham1206agra — APPROVED** [review]

**Review by cristipaval — APPROVED** [review]

**Comment by puneetlath** [comment]
Yes, that's right. Just like workspace chats we'd have the big sender avatar of the sender's workspace with the small payer user avatar.

**Comment by VickyStash** [comment]
@cristipaval Can we please create a follow-up issue to update the invoice room avatars display?
cc @waterim 

> Similar comment here as the other PR, I think we might not be using the correct room avatars here? cc @puneetlath
> 
> I think the idea was to do this for Workspace to Workspace: image
> 
> And to do this for Workspace to Individual: image
> 
> Does that sound right Puneet & @davidcardoza ?

**Comment by danielrvidal** [comment]
>Can we please create a follow-up issue to update the invoice room avatars display?
I created the issue here: https://github.com/Expensify/App/issues/41261

Let me know if I need to do anything else.

**Comment by MonilBhavsar** [comment]
There's a bug noted mostly coming from this PR https://github.com/Expensify/App/issues/41287. If anyone can please take a look

File: src/libs/API/parameters/SendInvoiceParams.ts
Comment (roryabraham) [inline src/libs/API/parameters/SendInvoiceParams.ts:4]: Nice little pro tip: _either_ `receiverEmail` _or_ `receiverInvoiceRoomID` is required. So I see why you made them both optional, but it later led to a problem [here](https://github.com/Expensify/App/issues/43060). So what you can do is use [RequireAtLeastOne](https://github.com/sindresorhus/type-fest/blob/main/source/require-at-least-one.d.ts) from type-fest.

**Comment by Pujan92** [comment]
We missed considering the enabled features of the collect workspace to show in the send invoice confirmation screen which raised an issue [here](https://github.com/Expensify/App/issues/41281).

File: src/libs/actions/Policy.ts
Comment (rojiphil) [inline src/libs/actions/Policy.ts:285]: As we did not exclude personal policies while fetching primary policy, this resulted in #46705

File: src/libs/actions/IOU.ts
Comment (mollfpr) [inline src/libs/actions/IOU.ts:1043]: ✋ Coming from https://github.com/Expensify/App/issues/54035

We already have a logic to delete the transaction if the transaction pending action is ADD when closing the error.

https://github.com/Expensify/App/blob/d56dfeca08a93d9b3a131c8ba17e4783758dad0e/src/components/ReportActionItem/MoneyRequestView.tsx#L458-L472

So, clear the `pendingAction` makes the report not deleted. We only need to clear the `pendingAction` on success.

File: src/libs/actions/IOU.ts
Comment (hoangzinh) [inline src/libs/actions/IOU.ts:1635]: We should multiply by "-1" here before building optimistic data for a transaction in case of sending invoice. Otherwise, it will cause calculating total amount to be incorrect when offline. More details here https://github.com/Expensify/App/issues/61050#issuecomment-2838217762

File: src/libs/actions/IOU.ts
Comment (rojiphil) [inline src/libs/actions/IOU.ts:3372]: Sending a new invoice did not scroll down the view as we missed passing the current account id instead of the receiver's account id. We fixed this in #63089

File: src/libs/ReportUtils.ts
Comment (ikevin127) [inline src/libs/ReportUtils.ts:2540]: This check was missing an additional variable `(isManager || isChatReportArchived)` which led to this issue:
- https://github.com/Expensify/App/issues/65338

we applied the missing variable and fixed the issue by not allowing editing fields of an archived report.

File: src/libs/actions/IOU.ts
Comment (Krishna2323) [inline src/libs/actions/IOU.ts:3365]: This line caused #97203. `receiverEmail` was sent as a raw phone number, never normalized with `addSMSDomainIfPhoneNumber()` the way every sibling flow (`SendMoney`, `Split`, `PerDiem`, `MoneyRequestBuilder`) does. Backend expects `+<number>@expensify.sms`, rejects the raw number with `Invalid invoice receiver`, and the optimistic invoice room is left stuck on its skeleton loader. 

Fixed in #100892 by routing through the already-normalized `receiverLogin`.

**Pull request merged**
